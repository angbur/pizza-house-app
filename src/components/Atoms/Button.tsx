import clsx from 'clsx';
import { CSSProperties, MouseEventHandler, PropsWithChildren } from 'react';
import styled from 'styled-components';

export const ButtonStyled = styled.div`
    background: #EB853E;
    padding: 1.25rem 0.625rem;
`

export type ButtonProps = {
  variant?: string;
  size?: string;
  className?: string;
  style?: CSSProperties;
  onClick?: MouseEventHandler<HTMLButtonElement>;
};

const Button = ({
  variant,
  size,
  className,
  style,
  children,
  onClick,
}: PropsWithChildren<ButtonProps>) => {
  return (
    <button className={clsx(variant, size, className)} onClick={onClick} style={style}>
      {children}
    </button>
  );
};

export default Button;
