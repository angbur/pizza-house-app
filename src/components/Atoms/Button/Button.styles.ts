import styled from 'styled-components';

const BaseButton = styled.button`
  text-transform: uppercase;
  font-weight: 600;
  letter-spacing: 1px;
  line-height: 1.4;
  border-radius: 8px;
  cursor: pointer;
  background: none;
  border: none;
  transition: 3ms;
`;

export const PrimaryLightButton = styled(BaseButton)`
  background: ${(props) => props.theme?.palette.primary};
  border: 1px solid ${(props) => props.theme?.palette.primary};
  color: ${(props) => props.theme?.palette.light};
  &:hover {
    background: ${(props) => props.theme?.palette.primaryDark};
  }
`;

export const PrimaryDarkButton = styled(BaseButton)`
  background: ${(props) => props.theme?.palette.secondary};
  border: 2px solid ${(props) => props.theme?.palette.secondary};
  color: ${(props) => props.theme?.palette.light};
  &:hover {
    background: ${(props) => props.theme?.palette.light};
    color: ${(props) => props.theme?.palette.secondary};
    border: 2px solid ${(props) => props.theme?.palette.secondary};
  }
`;

export const SecondaryLightButton = styled(BaseButton)`
  border: 2px solid ${(props) => props.theme?.palette.primary};
  color: ${(props) => props.theme?.palette.primary};
  &:hover {
    background: ${(props) => props.theme?.palette.primary};
    color: ${(props) => props.theme?.palette.secondary};
  }
`;

export const SecondaryDarkButton = styled(BaseButton)`
  border: 2px solid ${(props) => props.theme?.palette.secondary};
  color: ${(props) => props.theme?.palette.secondary};
  &:hover {
    background: ${(props) => props.theme?.palette.primary};
    border: 2px solid ${(props) => props.theme?.palette.primary};
  }
`;

const BaseTextButton = styled(BaseButton)`
  position: relative;
  display: inline-block;
  &::after {
    content: '';
    position: absolute;
    width: 100%;
    transform: scaleX(0);
    height: 2px;
    bottom: 0;
    left: 0;
    transform-origin: bottom right;
    transition: transform 0.25s ease-out;
  }
  &:hover {
    &::after {
      transform: scaleX(1);
      transform-origin: bottom left;
    }
  }
`;

export const TextLightButton = styled(BaseTextButton)`
  color: ${(props) => props.theme?.palette.primary};
  &::after {
    background-color: ${(props) => props.theme?.palette.primary};
  }
`;

export const TextDarkButton = styled(BaseTextButton)`
  color: ${(props) => props.theme?.palette.secondary};
  &::after {
    background-color: ${(props) => props.theme?.palette.secondary};
  }
`;

export const IconButton = styled(BaseButton)`
  color: ${(props) => props.theme?.palette.secondary};
  scale: 200%;
`;
