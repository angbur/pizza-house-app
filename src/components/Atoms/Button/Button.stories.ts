import { getInitialTheme } from './../../Theme/initialTheme';
import { Meta, StoryObj } from '@storybook/react';
import Button, { buttonSizes, buttonVariants } from './Button';

const meta = {
  title: 'Atoms/Button',
  component: Button,
  tags: ['docsPage'],
  argTypes: {
    variant: buttonVariants,
    size: {
      options: buttonSizes,
      control: { type: 'radio' },
    },
    children: {
      defaultValue: 'Press me',
    },
    theme: {
      defaultValue: getInitialTheme()
    }
  },
} satisfies Meta<typeof Button>;

export default meta;

type Story = StoryObj<typeof meta>;

export const PrimaryLight: Story = {
  args: {
    variant: 'primary-light',
  },
};

export const PrimaryDark: Story = {
  args: {
    variant: 'primary-dark',
  },
};

export const SecondaryLight: Story = {
  args: {
    variant: 'secondary-light',
  },
};

export const ButtonTextLight: Story = {
  args: {
    variant: 'button-text-light',
  },
};

export const ButtonTextDark: Story = {
  args: {
    variant: 'button-text-dark',
  },
};
