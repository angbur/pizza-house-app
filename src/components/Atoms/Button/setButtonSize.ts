import { CSSProperties } from 'react';
import { ButtonProps } from './Button';

const setButtonSize = ({ size, style }: ButtonProps): CSSProperties => {
  let scale = 1;

  if (size === 'md') scale = 0.9;
  if (size === 'sm') scale = 0.8;

  const padding: CSSProperties = {
    padding: `${scale * 0.625}rem ${scale * 1.25}rem`,
    fontSize: `${(scale + 0.05) * 1}rem`,
  };

  return {
    ...padding,
    ...style,
  };
};

export default setButtonSize;