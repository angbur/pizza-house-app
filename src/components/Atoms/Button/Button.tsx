import clsx from 'clsx';
import {
  ComponentPropsWithoutRef,
  CSSProperties,
  MouseEventHandler,
  PropsWithChildren,
  useContext,
} from 'react';
import { ThemeContext } from 'components/Theme/ThemeContext';
import {
  IconButton,
  PrimaryDarkButton,
  PrimaryLightButton,
  SecondaryDarkButton,
  SecondaryLightButton,
  TextDarkButton,
  TextLightButton,
} from './Button.styles';
import setButtonSize from './setButtonSize';

export const buttonVariants = [
  'primary-light',
  'primary-dark',
  'secondary-light',
  'secondary-dark',
  'button-text-light',
  'button-text-dark',
  'icon-button',
] as const;

type ButtonVariant = typeof buttonVariants[number];
export const buttonSizes = ['sm', 'md', 'lg'] as const;
export type ButtonSize = typeof buttonSizes[number];

export type ButtonProps = {
  variant?: ButtonVariant;
  size?: ButtonSize;
  className?: string;
  style?: CSSProperties;
  children?: PropsWithChildren;
  onClick?: MouseEventHandler<HTMLButtonElement>;
};

const Button = ({
  variant = 'primary-light',
  size = 'lg',
  className,
  style,
  children,
  onClick,
}: PropsWithChildren<ButtonProps> & ComponentPropsWithoutRef<'button'>) => {
  const theme = useContext(ThemeContext);

  const props = {
    className: clsx(variant, `btn-${size}`, className, 'button'),
    onClick: onClick,
    style: { ...setButtonSize(size), ...style },
    theme: theme,
  };

  return (
    <>
      {variant === 'primary-light' ? (
        <PrimaryLightButton {...props}>{children}</PrimaryLightButton>
      ) : variant === 'primary-dark' ? (
        <PrimaryDarkButton {...props}>{children}</PrimaryDarkButton>
      ) : variant === 'secondary-light' ? (
        <SecondaryLightButton {...props}>{children}</SecondaryLightButton>
      ) : variant === 'secondary-dark' ? (
        <SecondaryDarkButton {...props}>{children}</SecondaryDarkButton>
      ) : variant === 'button-text-light' ? (
        <TextLightButton {...props}>{children}</TextLightButton>
      ) : variant === 'button-text-dark' ? (
        <TextDarkButton {...props}>{children}</TextDarkButton>
      ) : variant === 'icon-button' ? (
        <IconButton {...props}>{children}</IconButton>
      ) : (
        <PrimaryLightButton {...props}>{children}</PrimaryLightButton>
      )}
    </>
  );
};

export default Button;
