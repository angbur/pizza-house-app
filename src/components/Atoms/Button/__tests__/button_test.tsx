import Button, { ButtonSize, ButtonVariant } from 'components/Atoms/Button/Button';
import { renderWithProviders } from 'test/test-utils';

type WrapperProps = {
  size?: ButtonSize;
  variant?: ButtonVariant;
};

const button = ({ size, variant }: WrapperProps) => {
  return renderWithProviders(<Button size={size} variant={variant} />, {
    preloadedState: {},
  }).container;
};

describe('Button', () => {
  test('renders correctly small size button', () => {
    expect(button({ size: 'sm' })).toMatchSnapshot();
  });

  test('renders correctly medium size button', () => {
    expect(button({ size: 'md' })).toMatchSnapshot();
  });

  test('renders correctly large size button', () => {
    expect(button({ size: 'lg' })).toMatchSnapshot();
  });

  test('renders correctly primary-light button', () => {
    expect(button({ variant: 'primary-light' })).toMatchSnapshot();
  });

  test('renders correctly primary-dark button', () => {
    expect(button({ variant: 'primary-dark' })).toMatchSnapshot();
  });

  test('renders correctly secondary-light button', () => {
    expect(button({ variant: 'secondary-light' })).toMatchSnapshot();
  });

  test('renders correctly secondary-dark button', () => {
    expect(button({ variant: 'secondary-dark' })).toMatchSnapshot();
  });

  test('renders correctly button text-dark', () => {
    expect(button({ variant: 'button-text-dark' })).toMatchSnapshot();
  });

  test('renders correctly button text-light', () => {
    expect(button({ variant: 'button-text-light' })).toMatchSnapshot();
  });

  test('renders correctly button text without hover', () => {
    expect(button({ variant: 'button-text-without-hover' })).toMatchSnapshot();
  });

  test('renders correctly icon button', () => {
    expect(button({ variant: 'icon-button' })).toMatchSnapshot();
  });
});
