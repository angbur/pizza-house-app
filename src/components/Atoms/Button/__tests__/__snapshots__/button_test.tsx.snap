// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly button text without hover 1`] = `
<div>
  <button
    class="sc-bcXHqe sc-jSUZER feNVH dkmCUq button-text-without-hover btn-lg button"
    style="padding: 0.65rem 1.25rem; font-size: 1.05rem;"
  />
</div>
`;

exports[`Button renders correctly button text-dark 1`] = `
<div>
  <button
    class="sc-bcXHqe sc-gKPRtg sc-ftTHYK feNVH hUzBlu hIdUJX button-text-dark btn-lg button"
    style="padding: 0.65rem 1.25rem; font-size: 1.05rem;"
  />
</div>
`;

exports[`Button renders correctly button text-light 1`] = `
<div>
  <button
    class="sc-bcXHqe sc-gKPRtg sc-iBYQkv feNVH hUzBlu YZIpJ button-text-light btn-lg button"
    style="padding: 0.65rem 1.25rem; font-size: 1.05rem;"
  />
</div>
`;

exports[`Button renders correctly icon button 1`] = `
<div>
  <button
    class="sc-bcXHqe sc-pyfCe feNVH hIOFrL icon-button btn-lg button"
    style="padding: 0.65rem 1.25rem; font-size: 1.05rem;"
  />
</div>
`;

exports[`Button renders correctly large size button 1`] = `
<div>
  <button
    class="sc-bcXHqe sc-gswNZR feNVH gHullA primary-light btn-lg button"
    style="padding: 0.65rem 1.25rem; font-size: 1.05rem;"
  />
</div>
`;

exports[`Button renders correctly medium size button 1`] = `
<div>
  <button
    class="sc-bcXHqe sc-gswNZR feNVH gHullA primary-light btn-md button"
    style="padding: 0.5850000000000001rem 1.125rem; font-size: 0.9500000000000001rem;"
  />
</div>
`;

exports[`Button renders correctly primary-dark button 1`] = `
<div>
  <button
    class="sc-bcXHqe sc-dkrFOg feNVH krmmpI primary-dark btn-lg button"
    style="padding: 0.65rem 1.25rem; font-size: 1.05rem;"
  />
</div>
`;

exports[`Button renders correctly primary-light button 1`] = `
<div>
  <button
    class="sc-bcXHqe sc-gswNZR feNVH gHullA primary-light btn-lg button"
    style="padding: 0.65rem 1.25rem; font-size: 1.05rem;"
  />
</div>
`;

exports[`Button renders correctly secondary-dark button 1`] = `
<div>
  <button
    class="sc-bcXHqe sc-eDvSVe feNVH cAJNIx secondary-dark btn-lg button"
    style="padding: 0.65rem 1.25rem; font-size: 1.05rem;"
  />
</div>
`;

exports[`Button renders correctly secondary-light button 1`] = `
<div>
  <button
    class="sc-bcXHqe sc-hLBbgP feNVH joABVS secondary-light btn-lg button"
    style="padding: 0.65rem 1.25rem; font-size: 1.05rem;"
  />
</div>
`;

exports[`Button renders correctly small size button 1`] = `
<div>
  <button
    class="sc-bcXHqe sc-gswNZR feNVH gHullA primary-light btn-sm button"
    style="padding: 0.52rem 1rem; font-size: 0.8500000000000001rem;"
  />
</div>
`;
